File: src/main/java/rag/trace/TraceEvent.java
package main.java.rag.trace;

/**
 * Represents a trace event for a pipeline stage.
 * Contains stage name, inputs, outputs summary, timing, and optional error.
 */
public class TraceEvent {
    private String stage;
    private String inputs;
    private String outputsSummary;
    private long timingMs;
    private String error;
    
    public TraceEvent(String stage, String inputs, String outputsSummary, long timingMs) {
        this.stage = stage;
        this.inputs = inputs;
        this.outputsSummary = outputsSummary;
        this.timingMs = timingMs;
        this.error = null;
    }
    
    public TraceEvent(String stage, String inputs, String outputsSummary, long timingMs, String error) {
        this.stage = stage;
        this.inputs = inputs;
        this.outputsSummary = outputsSummary;
        this.timingMs = timingMs;
        this.error = error;
    }
    
    public String getStage() {
        return stage;
    }
    
    public String getInputs() {
        return inputs;
    }
    
    public String getOutputsSummary() {
        return outputsSummary;
    }
    
    public long getTimingMs() {
        return timingMs;
    }
    
    public String getError() {
        return error;
    }
    
    public boolean hasError() {
        return error != null;
    }
    
    /**
     * Convert to JSON string for JSONL format.
     */
    public String toJson() {
        StringBuilder json = new StringBuilder();
        json.append("{");
        json.append("\"stage\":\"").append(escapeJson(stage)).append("\",");
        json.append("\"inputs\":\"").append(escapeJson(inputs)).append("\",");
        json.append("\"outputsSummary\":\"").append(escapeJson(outputsSummary)).append("\",");
        json.append("\"timingMs\":").append(timingMs);
        if (error != null) {
            json.append(",\"error\":\"").append(escapeJson(error)).append("\"");
        }
        json.append("}");
        return json.toString();
    }
    
    private String escapeJson(String str) {
        if (str == null) return "";
        return str.replace("\\", "\\\\")
                  .replace("\"", "\\\"")
                  .replace("\n", "\\n")
                  .replace("\r", "\\r")
                  .replace("\t", "\\t");
    }
}

File: iteration1_py/src/trace/trace_event.py
class TraceEvent:

    def __init__(self, stage: str, inputs: str, outputs_summary: str, timing_ms: int, error: str = None):
        self.__stage = stage
        self.__inputs = inputs
        self.__outputs_summary = outputs_summary
        self.__timing_ms = timing_ms
        self.__error = error

    # Getter methods
    def get_stage(self) -> str:
        return self.__stage

    def get_inputs(self) -> str:
        return self.__inputs

    def get_outputs_summary(self) -> str:
        return self.__outputs_summary

    def get_timing_ms(self) -> int:
        return self.__timing_ms

    def get_error(self) -> str:
        return self.__error

    def has_error(self) -> bool:
        return self.__error is not None

    def to_json(self) -> str:
        json_parts = [
            f'"stage":"{self.__escape_json(self.__stage)}"',
            f'"inputs":"{self.__escape_json(self.__inputs)}"',
            f'"outputsSummary":"{self.__escape_json(self.__outputs_summary)}"',
            f'"timingMs":{self.__timing_ms}'
        ]
        if self.__error is not None:
            json_parts.append(f'"error":"{self.__escape_json(self.__error)}"')
        return "{" + ",".join(json_parts) + "}"

    # Private helper method
    def __escape_json(self, s: str) -> str:
        if s is None:
            return ""
        return s.replace("\\", "\\\\") \
                .replace("\"", "\\\"") \
                .replace("\n", "\\n") \
                .replace("\r", "\\r") \
                .replace("\t", "\\t")
